akka-stream-tests/src/test/scala/akka/stream/impl/fusing/KeepGoingStageSpec.scala

```diff
@@ -0,0 +1,201 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl.fusing
+
+import akka.actor.{ NoSerializationVerificationNeeded, ActorRef }
+import akka.stream.scaladsl.{ Keep, Source, Sink }
+import akka.stream.{ Attributes, Inlet, SinkShape, ActorMaterializer }
+import akka.stream.stage.{ InHandler, AsyncCallback, GraphStageLogic, GraphStageWithMaterializedValue }
+import akka.stream.testkit.AkkaSpec
+import akka.stream.testkit.Utils._
+
+import scala.concurrent.{ Await, Promise, Future }
+import scala.concurrent.duration._
+
+class KeepGoingStageSpec extends AkkaSpec {
+
+  implicit val mat = ActorMaterializer()
+
+  trait PingCmd extends NoSerializationVerificationNeeded
+  case class Register(probe: ActorRef) extends PingCmd
+  case object Ping extends PingCmd
+  case object CompleteStage extends PingCmd
+  case object FailStage extends PingCmd
+  case object Throw extends PingCmd
+
+  trait PingEvt extends NoSerializationVerificationNeeded
+  case object Pong extends PingEvt
+  case object PostStop extends PingEvt
+  case object UpstreamCompleted extends PingEvt
+  case object EndOfEventHandler extends PingEvt
+
+  case class PingRef(private val cb: AsyncCallback[PingCmd]) {
+    def register(probe: ActorRef): Unit = cb.invoke(Register(probe))
+    def ping(): Unit = cb.invoke(Ping)
+    def stop(): Unit = cb.invoke(CompleteStage)
+    def fail(): Unit = cb.invoke(FailStage)
+    def throwEx(): Unit = cb.invoke(Throw)
+  }
+
+  class PingableSink(keepAlive: Boolean) extends GraphStageWithMaterializedValue[SinkShape[Int], Future[PingRef]] {
+    val shape = SinkShape[Int](Inlet("ping.in"))
+
+    override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[PingRef]) = {
+      val promise = Promise[PingRef]()
+
+      val logic = new GraphStageLogic(shape) {
+        private var listener: Option[ActorRef] = None
+
+        override def preStart(): Unit = {
+          promise.trySuccess(PingRef(getAsyncCallback(onCommand)))
+        }
+
+        private def onCommand(cmd: PingCmd): Unit = cmd match {
+          case Register(probe) ⇒ listener = Some(probe)
+          case Ping            ⇒ listener.foreach(_ ! Pong)
+          case CompleteStage ⇒
+            completeStage()
+            listener.foreach(_ ! EndOfEventHandler)
+          case FailStage ⇒
+            failStage(TE("test"))
+            listener.foreach(_ ! EndOfEventHandler)
+          case Throw ⇒
+            try {
+              throw TE("test")
+            } finally listener.foreach(_ ! EndOfEventHandler)
+        }
+
+        setHandler(shape.inlet, new InHandler {
+          override def onPush(): Unit = pull(shape.inlet)
+
+          // Ignore finish
+          override def onUpstreamFinish(): Unit = listener.foreach(_ ! UpstreamCompleted)
+        })
+
+        override def keepGoingAfterAllPortsClosed: Boolean = keepAlive
```

Comment:
bikeshed: `keepAliveAfter...` sounds a bit more like what it is, but we have `keepAlive` as method so to avoid confusion maybe keepGoing is OK... (thinking out loud, no need to change I guess)